lib/puppetserver/ca/action/prune.rb

```diff
@@ -0,0 +1,149 @@
+              parsed['config'] = conf
+            end
+          end
+        end
+
+        def run(inputs)
+          config_path = inputs['config']
+
+          # Validate the config path.
+          if config_path
+            errors = FileSystem.validate_file_paths(config_path)
+            return 1 if Errors.handle_with_usage(@logger, errors)
+          end
+
+          # Validate puppet config setting.
+          settings_overrides = {}
+          puppet = Config::Puppet.new(config_path)
+          puppet.load(settings_overrides)
+          return 1 if Errors.handle_with_usage(@logger, puppet.errors)
+
+          # Validate that we are offline
+          return 1 if HttpClient.check_server_online(puppet.settings, @logger)
+
+          # Getting the CRL(s)
+          loader = X509Loader.new(puppet.settings[:cacert], puppet.settings[:cakey], puppet.settings[:cacrl])
+
+          # Take 'em
+          puppet_crl = select_puppet_CRL(loader.crls, loader.key)
+          # Prune 'em
+          prune_CRLs(puppet_crl)
+          # Update 'em
+          update_pruned_CRL(puppet_crl, loader.key)
+          # Write 'em out
+          write_crl_to_file(loader.crls, puppet.settings[:cacrl])
+
+          @logger.inform("Finished pruning Puppet's CRL")
+          return 0
+        end
+
+        def select_puppet_CRL(crl_array,key)
+          puppet_crl = Array.new
+          crl_array.each do |crl|
+            if crl.verify(key)
+              puppet_crl << crl
+            end
+          end
+          return puppet_crl
+        end
+
+        def prune_CRLs(crl_list)
+          crl_list.each do |list|
+            existed_serial_number = Set.new()
+            revoked_list = list.revoked
+            puts("Issuer #{list.issuer.to_s(OpenSSL::X509::Name::RFC2253)}") if @logger.debug?
+
+            revoked_list.delete_if do |revoked|
+              if existed_serial_number.add?(revoked.serial)
+                false
+              else
+                debug_output(revoked.serial, revoked.time) if @logger.debug?
+                true
+              end
+            end
+            list.revoked=(revoked_list)
+          end
+        end
+
+        def update_pruned_CRL(crl_list, pkey)
+          crl_list.each do |crl|
+            crl.version=(crl.version + 1)
+            crl.sign(pkey, OpenSSL::Digest::SHA256.new)
+          end
+        end
+
+        def write_crl_to_file(crl_to_write, file_path)
+          begin
+            file = File.open(file_path, "w")
+            crl_to_write.each do |crl|
+              file.write(crl.to_pem)
+            end
+          rescue IOError => e
+            @logger.err("Error when write out to file on disk")
+            @logger.err("Error: " + e.message)
+          ensure
+            file.close unless file.nil?
+          end
+        end
+
+        # Output for debug
+        #  - Serial number and time of revocation
+        def debug_output(serial,time)
```

Comment:
Not sure this needs to be its own method, since it's only used once and it's literally just logging a message.
It's easier to read the `prune_CRLs` method and see what it's logging if you don't have to scroll to a different method.